src/EditorFeatures/Core/Implementation/IntelliSense/Completion/AsyncCompletionService.cs

```diff
@@ -87,10 +87,32 @@ private bool UseLegacyCompletion(ITextView textView, ITextBuffer subjectBuffer)
         {
             if (!_newCompletionAPIEnabled.HasValue)
             {
-                if (Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out var workspace))
+                int userSetting = 0;
+                const string useAsyncCompletionOptionName = "UseAsyncCompletion";
+                if (textView.Options.GlobalOptions.IsOptionDefined(useAsyncCompletionOptionName, localScopeOnly: false))
                 {
-                    var experimentationService = workspace.Services.GetService<IExperimentationService>();
-                    _newCompletionAPIEnabled = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.CompletionAPI);
+                    userSetting = textView.Options.GlobalOptions.GetOptionValue<int>(useAsyncCompletionOptionName);
+                }
+
+                // The meaning of the UseAsyncCompletion option definition's values:
+                // -1 - user disabled async completion
+                //  0 - no changes from the user; check the experimentation service for whether to use async completion
+                //  1 - user enabled async completion
+                if (userSetting == 1)
+                {
+                    _newCompletionAPIEnabled = true;
+                }
+                else if (userSetting == -1)
+                {
+                    _newCompletionAPIEnabled = false;
+                }
+                else
+                {
+                    if (Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out var workspace))
+                    {
+                        var experimentationService = workspace.Services.GetService<IExperimentationService>();
+                        _newCompletionAPIEnabled = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.CompletionAPI);
```

Comment:
if there is a chance of the if statement failing, consider setting `_newCompletionAPIEnabled` to something so that we avoid repeating work. If this won't fail, though, don't worry about it.

Reply:
As soon as we get a workspace and an experimentation service, we should set _newCompletionAPIEnabled to a value. We should try checking unit this. Not?

---
In reply to: [245185354](https://github.com/dotnet/roslyn/pull/32045#discussion_r245185354) [](ancestors = 245185354)